code/datums/components/nanites.dm

```diff
@@ -200,12 +200,27 @@
 		var/datum/nanite_program/NP = X
 		NP.on_emp(severity)
 
-/datum/component/nanites/proc/on_shock(datum/source, shock_damage)
-	nanite_volume *= (rand(45, 80) * 0.01)		//Lose 20-55% of nanites
-	adjust_nanites(null, -(rand(5, 50)))			//Lose 5-50 flat nanite volume
-	for(var/X in programs)
-		var/datum/nanite_program/NP = X
-		NP.on_shock(shock_damage)
+
+/datum/component/nanites/proc/on_shock(datum/source, shock_damage, siemens_coeff = 1, flags = NONE)
+	if(flags & SHOCK_ILLUSION || shock_damage < 1)
+		return
+
+	var/has_no_nanite_shockimune = FALSE
+	if(HAS_TRAIT_FROM(host_mob, TRAIT_SHOCKIMMUNE, "nanites"))
+		REMOVE_TRAIT(host_mob, TRAIT_SHOCKIMMUNE, "nanites")
+		if(HAS_TRAIT(host_mob, TRAIT_SHOCKIMMUNE))
+			has_no_nanite_shockimune = TRUE
+		ADD_TRAIT(host_mob, TRAIT_SHOCKIMMUNE, "nanites")

```

Comment:
my god

Reply:
If you have better way to check is there no nanite shockimmune, help me please.

Reply:
`#define HAS_TRAIT_NOT_FROM(target, trait, source) (target.status_traits ? (target.status_traits[trait] ? (length(target.status_traits[trait] - source) > 0) : FALSE) : FALSE)`
add this to the defines file.

This should, in theory (can't test it right now), return TRUE if the target has the trait from any source but the one you give it. It could probably also be extended to accept a list of multiple sources to exclude, but i'm not sure if it could cause complications as a define.

Reply:
Thanks! I am sure that this will work with the list of traits.
I test it some.